### PR TITLE
Split historic forecasts and truths based on metadata for EMOS CLIs

### PR DESCRIPTION
**Pull request by gavinevans**
Description
Addition of a new plugin with associated changes to CLIs so that the historic forecasts and truths can be supplied to the CLIs as a single argument, which is then separated into the historic forecasts and truths based on the metadata.

The SplitHistoricForecastAndTruth plugin allows the historic forecasts and truths to be passed to the `estimate-emos-coefficients` CLI as a single string of filepaths containing both historic forecasts and truth. This plugin then splits the filepaths into the historic forecasts and truth based on the metadata provided. The modification to add this plugin has been pulled through into the `estimate-emos-coefficients` and `ensemble-calibration` CLIs with most of the files changed within this PR being CLI tests associated with these CLI changes.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 1]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+                    A cube containing the truth datasets.
+        """
+        historic_forecasts = self._find_required_cubes_using_metadata(
+            cubes, self.historic_forecast_dict)
+        truths = self._find_required_cubes_using_metadata(
+            cubes, self.truth_dict)
+
+        if len(historic_forecasts) != len(truths):
+            historic_forecasts, truths = (
+                self._filter_non_matching_cubes(historic_forecasts, truths))
```
Comment (bayliffe) [inline lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: Should we apply this filtering even if the two are the same length? That would avoid any hidden mismatching, e.g.:

I have 15 days worth of data looking at T+10. I am missing the T+10 forecast from day 1, and I am missing the truth on day 5. I still get 14 historic forecasts and 14 truths, they are just mismatched.


In the same vein, do we need to sort these lists? Is there some matching of validity times in the ensemble calibration coefficient calculation to ensure we are looking at equivalent validity times between forecast and truth (I assume so, but could not find it at first glance)?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        self.assertEqual(truth_result, self.partial_truth)
+
+    def test_fewer_truths(self):
+        """Test for when there are fewer truths than historic forecasts,
+        for example, if there is a missing analysis."""
+        hf_result, truth_result = self.plugin._filter_non_matching_cubes(
+            self.historic_forecasts, self.partial_truth)
+        self.assertEqual(hf_result, self.partial_historic_forecasts)
+        self.assertEqual(truth_result, self.partial_truth)
+
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: In line with my comment about the call to this function, could you also have a test where they are mismatched, but differently so (e.g. one missing the first validity time, and one the last, resulting in a list that is 2 elements shorter than the total possible number).

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    def test_fewer_truths(self):
+        """Test that the input cubelist combining historic forecasts and truth
+        can be split using the metadata dictionaries provided, when there are
+        fewer truths than historic forecasts."""
+        combined = self.historic_forecasts + self.partial_truth
+
+        hf_result, truth_result = self.plugin.process(combined)
+        self.assertEqual(hf_result, self.partial_historic_forecasts_cube)
+        self.assertEqual(truth_result, self.partial_truth_cube)
+
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: Likewise, if you accept my point about it being worth always applying the filter, we should have an offset mismatch test here as to ensure it invokes the filtering.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        self.partial_historic_forecasts_cube = (
+            self.partial_historic_forecasts.merge_cube())
+        self.partial_truth = self.truth[:2] + self.truth[3:]
+        self.partial_truth_cube = self.partial_truth.merge_cube()
+
+        # Set up dictionaries specifying the metadata to identify the
+        # historical forecasts and truth
+        self.historic_forecast_dict = {
+            "attributes": {
+                "mosg__model_configuration": "uk_ens"
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: Just to let you know, it came up today that using mosg__ attributes aren't very future proof as they aren't very future proof. At some point we may need to add some kind of "truth_or_forecast" attribute to the cubes as they are stored in the EMOS cache and use this EMOS specific attribute.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 5]
```diff
@@ -107,18 +125,39 @@ def main(argv=None):
     args = parser.parse_args(args=argv)
 
     # Load Cubes
-    historic_forecast = load_cube(args.historic_filepath)
-    truth = load_cube(args.truth_filepath)
+    historic_forecast = load_cube_or_none(args.historic_filepath)
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: FYI this function may be disappearing in @neilCrosswaite's work.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 6]
```diff
@@ -69,12 +74,25 @@ def main(argv=None):
                              'This cycletime is in the format '
                              'YYYYMMDDTHHMMZ.')
     # Filepaths for historic and truth data.
-    parser.add_argument('historic_filepath', metavar='HISTORIC_FILEPATH',
+    parser.add_argument('--historic_filepath', metavar='HISTORIC_FILEPATH',
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: It's not necessarily clear that these options are mutually exclusive, you want either (--historic_filepath and --truth_filepath) or --combined_filepath.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 7]
```diff
@@ -69,12 +74,25 @@ def main(argv=None):
...
                         help='A path to an input NetCDF file containing the '
                              'historic forecast(s) used for calibration.')
-    parser.add_argument('truth_filepath', metavar='TRUTH_FILEPATH',
+    parser.add_argument('--truth_filepath', metavar='TRUTH_FILEPATH',
                         help='A path to an input NetCDF file containing the '
                              'historic truth analyses used for calibration.')
+    # Input filepaths
+    parser.add_argument('--combined_filepath', metavar='COMBINED_FILEPATH',
+                        help='The path to the input NetCDF files containing '
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: This is actually a list of several path**s**? Should it have an nargs='+' type thing going on? Same for the truth and historic filepath arguements.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 8]
```diff
@@ -107,18 +125,39 @@ def main(argv=None):
...
+               "the historic_filepath {} and the truth_filepath {} "
+               "should be specified, or the combined_filepath {} should be "
+               "specified alongside the historic_forecast_identifier {} and "
+               "truth_identifier {}. In this case the arguments provided "
+               "were not sufficient.".format(
+                    args.historic_filepath, args.truth_filepath,
+                    args.combined_filepath, args.historic_forecast_identifier,
+                    args.truth_identifier))
+        warnings.warn(msg)
+        return
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/cli/estimate_emos_coefficients.py]: We should also probably check that we haven't supplied all of [historic_forecast, truth, combined] although we used to have some kind of mutually exclusive helper code for argparser which could be used here if we still have it.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 9]
```diff
@@ -169,6 +227,10 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
             contains a coefficient_index dimension coordinate and a
             coefficient_name auxiliary coordinate.
     """
+    if combined is not None:
+        historic_forecast, truth = SplitHistoricForecastAndTruth(
+            historic_forecast_dict, truth_dict).process(combined)
```
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/cli/estimate_emos_coefficients.py]: I think this could be solved earlier with the mutually exclusive argparser functionally, but maybe we need some guidance on CLI structure here too.

File: lib/improver/cli/ensemble_calibration.py [hunk 10]
```diff
@@ -172,14 +191,33 @@ def main(argv=None):
 
     # Load cubes
```
Comment (fionaRust) [inline lib/improver/cli/ensemble_calibration.py]: Is it worth considering factoring out any of this shared code, as we have basically the same stuff in both CLIs? Or is it not worth worrying about now?

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 5 above)
Comment (bayliffe) [reply lib/improver/cli/estimate_emos_coefficients.py]: It is disappearing as of #941 which may well be merged before Gavin returns. So yes, load_cube_or_none, if you want the return None behaviour is now:
```
load_cube(filepath, allow_none=True)
```

File: lib/improver/cli/ensemble_calibration.py [hunk 11]
```diff
@@ -325,6 +383,12 @@ def process(current_forecast, historic_forecast, truth, units, distribution,
         num_realizations = len(
             current_forecast.coord('realization').points)
 
+    # Separate the historic forecasts and the associated truths, if these
+    # have been supplied as a single input.
+    if combined is not None:
+        historic_forecast, truth = SplitHistoricForecastAndTruth(
+            historic_forecast_dict, truth_dict).process(combined)
+
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/ensemble_calibration.py]: I've now removed this CLI, as part of #945.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've moved this check to be within the `process` method and added some additional checks for the inputs to make sure that errors are raised, if invalid combinations are supplied.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added some checking of the inputs to avoid this.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 12]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+    @staticmethod
+    def _find_required_cubes_using_metadata(cubes, input_dict):
+        """
+        Extract the cube that matches the information within the input
+        dictionary.
+
+        Args:
+            cubes (iris.cube.CubeList):
+                The cube that will be checked to see whether the metadata
+                within this cube matches the input dictionary.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: Modified.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 13]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+            input_dict (dict):
+                A dictionary containing the metadata that will be used to
+                identify the desired cube.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: Corrected.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 14]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+
+        Returns:
+            iris.cube.CubeList:
+                CubeList containing cubes that matches the metadata supplied
+                within the input dictionary.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: Corrected.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py (hunk 1 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: After a conversation with @rust, I've actually moved the filtering function out of this plugin and into the `EstimateCoefficientsForEnsembleCalibration` plugin. This is because as you've noted, there isn't otherwise, any checking for equivalent validity times between the forecast and truth, however, this checking is now provided by the `_filter_non_matching_cubes` function.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py (hunk 2 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: I've added a test that is similar to what you've described.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py (hunk 3 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: As I've now moved the filtering function to the `EstimateCoefficientsForEnsembleCalibration` plugin, I've added a test to the `process` method of this plugin.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 15]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+            truth = truths.extract(constr)
+            if truth:
+                matching_historic_forecasts.append(cube)
+                matching_truths.extend(truth)
+        return matching_historic_forecasts, matching_truths
+
+    def process(self, cubes):
+        """
+        Separate the input cubes into the historic_forecasts and truth based
+        on the metadata information supplied within the input dictionaries.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: The filtering has now been moved into the `EstimateCoefficientsForEnsembleCalibration` plugin. I've extended the help within the CLI to be explicit about which combination of arguments are expected.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 16]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+        self.truth_dict = truth_dict
+
+    def __repr__(self):
+        """Represent the plugin instance as a string."""
+        result = ('<SplitHistoricForecastsAndTruth: '
+                  'historic_forecast_dict={}, truth_dict={}>')
+        return result.format(self.historic_forecast_dict, self.truth_dict)
+
+    @staticmethod
+    def _find_required_cubes_using_metadata(cubes, input_dict):
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: For the moment, I've left this as it was.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 17]
```diff
@@ -95,3 +96,145 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+                Only constraining on attributes is supported.
+            ValueError:
+                The metadata supplied resulted in no matching cubes.
+
+        """
+        for key in input_dict.keys():
+            if key == "attributes":
+                constr = iris.AttributeConstraint(**input_dict[key])
+            else:
+                msg = ("At present, only constraining on attributes is "
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: I've moved this checking into the `__init__`, so the `_find_required_cubes_using_metadata` is a simplified a bit.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py (hunk 4 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: Thanks. For the moment, I'm using the attributes but this could be modified in the future, if the metadata conventions change.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 5 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: Yes, I've now updated this to use the `load_cube` function with the `allow_none` keyword argument.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 6 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: Yes, I've extended the help within the CLI to try to be clear about which combinations are valid.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 7 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added `nargs='+'` as suggested.

File: lib/improver/cli/ensemble_calibration.py (hunk 10 above)
Comment (gavinevans) [reply lib/improver/cli/ensemble_calibration.py]: I've removed this CLI in #945. This change is also within this PR.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 18]
```diff
@@ -107,18 +152,27 @@ def main(argv=None):
...
-    coefficients = process(historic_forecast, truth, args.distribution,
-                           args.cycletime, args.units, args.predictor_of_mean,
-                           args.max_iterations)
+    coefficients = process(historic_forecast, truth, combined,
+                           historic_forecast_dict, truth_dict,
+                           args.distribution, args.cycletime, args.units,
+                           args.predictor_of_mean, args.max_iterations)
     # Save Cube
-    save_netcdf(coefficients, args.output_filepath)
+    if coefficients:
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Might be worth expanding the comment above to say why this is an if statement now

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 19]
```diff
@@ -68,13 +73,53 @@ def main(argv=None):
...
+        "--historic_forecast_identifier",
+        metavar='HISTORIC_FORECAST_IDENTIFIER',
+        help='The path to a json file containing metadata '
+             'information that defines the historic forecast. '
+             'This must be supplied with both the combined_filepath and the '
+             'truth_identifier. Specification of either the historic_filepath'
+             'or the truth_filepath is invalid with this argument.')
+    parser.add_argument(
+        "--truth_identifier", metavar='TRUTH_IDENTIFIER',
+        help='The path to a json file containing metadata '
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Could link to the IMPROVER function where the format of this json file is defined? Is that what we've done for this kind of thing before?

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 20]
```diff
@@ -133,6 +187,29 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
             The cube containing the historical forecasts used for calibration.
         truth (iris.cube.Cube):
             The cube containing the truth used for calibration.
+        combined (iris.cube.CubeList):
+            A cubelist containing a combination of historic forecasts and
+            associated truths.
+        historic_forecast_dict (dict):
+            Dictionary specifying the metadata that defines the historic
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Again maybe we could link to the plugin that uses this dictionary. What do you think?

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 21]
```diff
@@ -169,6 +246,52 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
             contains a coefficient_index dimension coordinate and a
             coefficient_name auxiliary coordinate.
     """
+    # Support the initial cycles when using a training dataset, where there
+    # might not yet be any available historic forecasts and truth to calibrate
+    # with. In this instance, only a warning is raised, as this would be
+    # expected behaviour when no training dataset exists.
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: This can be removed

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 22]
```diff
@@ -169,6 +246,52 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+               "should be specified, or the combined_filepath {} should be "
+               "specified alongside the historic_forecast_identifier {} and "
+               "truth_identifier {}. In this case the arguments provided "
+               "were not sufficient.".format(
+                    historic_forecast, truth, combined,
+                    historic_forecast_dict, truth_dict))
+        warnings.warn(msg)
+        return
+
+    if any([historic_forecast, truth]):
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Can we have comments above each of these blocks of if statements?

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 23]
```diff
@@ -169,6 +246,52 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+        if all([historic_forecast, truth]):
+            if any([combined, historic_forecast_dict, truth_dict]):
+                msg = ("If the historic_filepath and truth_filepath arguments "
+                       "are specified then none of the the combined_filepath, "
+                       "historic_forecast_identifier and truth_identifier "
+                       "arguments should be specified.")
+                raise ValueError(msg)
+            else:
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Do you need this else?

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 24]
```diff
@@ -169,6 +246,52 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+        else:
+            msg = ("Both the historic_filepath and truth_filepath arguments "
+                   "should be specified if one of these arguments are "
+                   "specified.")
+            raise ValueError(msg)
+
+    if any([combined, historic_forecast_dict, truth_dict]):
+        if all([combined, historic_forecast_dict, truth_dict]):
+            pass
+        else:
```
Comment (fionaRust) [inline lib/improver/cli/estimate_emos_coefficients.py]: Again I think this can be written as just an if not an if else.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fionaRust) [reply lib/improver/cli/estimate_emos_coefficients.py]: I think this is ok now

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py [hunk 25]
```diff
@@ -94,3 +94,113 @@ def check_predictor_of_mean_flag(predictor_of_mean_flag):
...
+                The metadata supplied resulted in no matching cubes.
+
+        """
+        constr = iris.AttributeConstraint(**input_dict["attributes"])
+        cubelist = cubes.extract(constr)
+        if not cubelist:
+            msg = ("The metadata to identify the desired historic forecast or "
+                   "truth has found nothing matching the metadata information "
+                   "supplied: {}".format(input_dict))
+            raise ValueError(msg)
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: Need to think about if this is desired behaviour from a suite perspective. Should this be a warning?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py [hunk 26]
```diff
@@ -109,24 +112,37 @@ def setUp(self):
...
+        # Create a combined list of historic forecasts and truth
+        self.combined = self.historic_forecasts + self.truth
 
-        self.temperature_truth_cube = (
-            _create_truth(self.current_temperature_forecast_cube))
+        # Create the historic and truth cubes
+        self.historic_forecasts_cube = self.historic_forecasts.merge_cube()
+        self.truth_cube = self.truth.merge_cube()
+
+        # Create cubes with alternative names
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: Why is this needed?

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py [hunk 27]
```diff
@@ -357,59 +373,36 @@ def add_forecast_reference_time_and_forecast_period(
...
 
 
-def _create_historic_forecasts(cube, number_of_days=5):
+def _create_historic_forecasts(data, time_dt, frt_dt,
+                               standard_grid_metadata="uk_ens",
+                               number_of_days=5, **kwargs):
     """
     Function to create a set of pseudo historic forecast cubes, based on the
     input cube, and assuming that there will be one forecast per day at the
     same hour of the day.
```
Comment (fionaRust) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: Do these set up functions need an Args section in the docstring or not?

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 18 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: This has now been removed.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 19 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added a link to the CLI help.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 20 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added a link to the CLI help.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 21 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: This has been removed.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 22 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added some comments around these if statements.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 23 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've removed this.

File: lib/improver/ensemble_calibration/ensemble_calibration.py [hunk 28]
```diff
@@ -702,6 +702,44 @@ def compute_initial_guess(
...
+        matching_truths = iris.cube.CubeList([])
+        for hf_slice in historic_forecast.slices_over("time"):
+            coord_values = (
+                {"time": iris_time_to_datetime(hf_slice.coord("time"))})
+            constr = iris.Constraint(coord_values=coord_values)
+            truth_slice = truth.extract(constr)
+            if truth_slice:
+                matching_historic_forecasts.append(hf_slice)
+                matching_truths.append(truth_slice)
+        return (matching_historic_forecasts.merge_cube(),
```
Comment (fionaRust) [inline lib/improver/ensemble_calibration/ensemble_calibration.py]: From a suite perspective we need to be careful about what happens if one of these merges fail. I think?

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 24 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've converted this into an `if not`.

File: lib/improver/ensemble_calibration/ensemble_calibration_utilities.py (hunk 25 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration_utilities.py]: After discussing this, this exception has been kept but this exception is converted into a warning within the CLI, as there would be instances, for example for the first day when initially creating a training dataset, the truth datasets will accumulate, however, there won't be a valid historic forecast until one day after the initial cycle, as we're only using historic forecasts with the same cycle point within our calibration.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py (hunk 26 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: I've removed this.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py (hunk 27 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/helper_functions.py]: I've added fuller docstrings for these functions.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 28 above)
Comment (gavinevans) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: I've added the raising of an exception to this function, as you're correct that if the historic forecasts and truths do not match in validity time, then we won't be able to calculate any coefficients.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py (hunk 4 above)
Comment (fionaRust) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration_utilities/test_SplitHistoricForecastAndTruth.py]: Caroline S was going to make a note of this in the relevant metadata epic. It might be that we need to update this to use be able to use a coordinate instead at some point.

File: lib/improver/ensemble_calibration/ensemble_calibration.py (hunk 28 above)
Comment (fionaRust) [reply lib/improver/ensemble_calibration/ensemble_calibration.py]: I think the chances of this happening in a suite context is very slim, even in the spin up period, so I'm happy that this raises an error.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 29]
```diff
@@ -168,11 +259,68 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+
+    Raises:
+        ValueError: If the historic forecast and truth inputs are specified,
+            then the combined input, historic forecast dictionary and truth
+            dictionary should not be specified.
+        ValueError: If one of the historic forecast or truth inputs are
+            specified, then they should both be specified.
+        ValueError: All of the combined_filepath, historic_forecast_identifier
+            and truth_identifier arguments should be specified if one of the
+            arguments are specified.
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: I've added a warns section.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 30]
```diff
@@ -168,11 +259,68 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+                   "arguments should be specified if one of the arguments are "
+                   "specified.")
+            raise ValueError(msg)
+
+    try:
+        if combined is not None:
+            historic_forecast, truth = SplitHistoricForecastAndTruth(
+                historic_forecast_dict, truth_dict).process(combined)
+    except ValueError as err:
+        if str(err).startswith("The metadata to identify the desired"):
```
[the remark this answers is not in the document]
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: Comment added.

File: lib/improver/cli/estimate_emos_coefficients.py [hunk 31]
```diff
@@ -168,11 +259,75 @@ def process(historic_forecast, truth, distribution, cycletime, units=None,
...
+
     """
-    result = EstimateCoefficientsForEnsembleCalibration(
-        distribution, cycletime, desired_units=units,
-        predictor_of_mean_flag=predictor_of_mean,
-        max_iterations=max_iterations).process(historic_forecast, truth)
+    # The logic for the if statements below is:
+    # 1. Check whether either the historic_forecast or the truth exists.
+    # 2. Check that both the historic forecast or the truth exists, otherwise,
+    #    raise an error.
```
Comment (bayliffe) [inline lib/improver/cli/estimate_emos_coefficients.py]: ...forecast **and** the truth...

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 32]
```diff
@@ -622,8 +618,71 @@ def test_mean_predictor_estimate_coefficients_nans(self):
...
+        hf_result, truth_result = Plugin._filter_non_matching_cubes(
+            self.partial_historic_forecasts, partial_truth)
+        self.assertEqual(hf_result, expected_historical_forecasts)
+        self.assertEqual(truth_result, expected_truth)
+
+    def test_no_matches_exception(self):
+        """Test for when there is both a missing historic forecasts and a
+        missing truth at different validity times. This results in the
+        expected historic forecasts and the expected truths containing cubes
+        at three matching validity times."""
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: This doc-string is incorrect.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py [hunk 33]
```diff
@@ -685,6 +744,24 @@ def test_coefficient_values_for_gaussian_distribution(self):
         self.assertArrayEqual(
             result.coord("coefficient_name").points, self.coeff_names)
 
+    def test_coefficient_values_for_gaussian_distribution_mismatching_inputs(
+            self):
+        """Test for when there is both a missing historic forecasts and a
+        missing truth at different validity times. This results in the
+        expected historic forecasts and the expected truths containing cubes
+        at three matching validity times."""
```
Comment (bayliffe) [inline lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: This doc-string is incorrect.

File: tests/improver-estimate-emos-coefficients/07-only_truth_error.bats [hunk 34]
```diff
@@ -0,0 +1,45 @@
...
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "estimate-emos-coefficients for no valid inputs" {
```
Comment (bayliffe) [inline tests/improver-estimate-emos-coefficients/07-only_truth_error.bats]: This test could do with a better name, as could those below which currently share the same name.

File: lib/improver/cli/estimate_emos_coefficients.py (hunk 31 above)
Comment (gavinevans) [reply lib/improver/cli/estimate_emos_coefficients.py]: Corrected.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 32 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: I've updated this docstring.

File: lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py (hunk 33 above)
Comment (gavinevans) [reply lib/improver/tests/ensemble_calibration/ensemble_calibration/test_EstimateCoefficientsForEnsembleCalibration.py]: I've updated this docstring.

File: tests/improver-estimate-emos-coefficients/07-only_truth_error.bats (hunk 34 above)
Comment (gavinevans) [reply tests/improver-estimate-emos-coefficients/07-only_truth_error.bats]: I've updated the descriptions of the tests as suggested.
